### PR TITLE
Feature: Mythic Keystone dungeon support

### PR DESCRIPTION
**Pull request by tdaugaard**
Re-based and fixed the old PR on top of the new main branch and redid the commits to make more logical sense, also removed some stuff that really wasn't specific to this feature and thus shouldn't go in this PR.

I have done extensive testing on this branch when doing M+ and there's only one thing I have yet to figure out how to fix, and that is that _all_ dungeons _except_ Karazhan are almost exactly on point with the VideoSegment timing, but _both_ Upper and Lower Karazhan are almost always off by at least 7-10 seconds. So basically, a boss segment starts, but you're 7-10 seconds into the boss fight. I think it's okay to leave it like that for now, as you can just press arrow left one time to go back 30 seconds and watch from there, until I can figure out exactly why it's happening for that dungeon.

This will fix issue #43

### CODE REVIEW

File: src/main/logutils.ts
Comment (aza547) [inline src/main/logutils.ts:495]: can we spread these args over some new lines for readability

File: src/main/logutils.ts
Comment (aza547) [inline src/main/logutils.ts:537]: What about dungeons where bosses into boss is pulled. I'm almost certain this happens with some routes.

File: src/main/logutils.ts
Comment (aza547) [inline src/main/logutils.ts:699]: Not sure how I feel about this. What about horribly bad runs with like 20 deaths, is that all going to appear in the context menu? Might get a bit busy. What about hunter feign deaths which causes a UNIT_DIED event to fire?

File: src/renderer/Layout.tsx
Comment (aza547) [inline src/renderer/Layout.tsx:258]: some brackets would make this more readable

File: src/renderer/Layout.tsx
Comment (aza547) [inline src/renderer/Layout.tsx:261]: That's cool... and maybe is reason for having larger resolutions. Would like to do this for other categories eventually.

File: src/renderer/VideoButton.tsx
Comment (aza547) [inline src/renderer/VideoButton.tsx:51]: brackets please!

File: src/renderer/images.ts
Comment (aza547) [inline src/renderer/images.ts:41]: empty try block?

**Review by aza547 — CHANGES_REQUESTED** [review]
- I thought video files should be 200x100 to be in-line with what we have already. I thought having large files causes the video button to display a corner of them. That being said it looks OK, so maybe you made some CSS changes that I've yet to see. 

- I don't think I'm a huge fan of the loading screen images (the WoW logo and text can overlap our text which I think looks bad. 

- The text is a different font/size to other buttons still. Can we make it consistent? 

- Logging is pretty ugly for the final metadata event, can we split so it's not a one line JSON string but a properly formatted JSON? 

I had to rush a little as getting dragged to a birthday dinner so my review of the renderer/ files was a bit limited in depth. 

Obviously this is a huge bit of work and will be awesome function to have, so thanks again for all your efforts. We should also re-advertise (as many of requested) when this is in (and confirmed to be stable, probably).

File: src/renderer/VideoButton.tsx
Comment (aza547) [inline src/renderer/VideoButton.tsx:152]: semi colons on a few lines here and below

File: src/renderer/rendererutils.ts
Comment (aza547) [inline src/renderer/rendererutils.ts:35]: Why does this need special handling? It is still a boolean at the end of the day?

**Review by aza547 — COMMENTED** [review]
What about timestamps in the test log (I commented on this in the other PR)?

File: src/main/constants.ts
Comment (tdaugaard) [reply src/main/constants.ts:151]: There was missing an 'n' from the end. 🤣

File: src/renderer/Layout.tsx
Comment (tdaugaard) [reply src/renderer/Layout.tsx:261]: Yeah, the larger images take up quite a bit of space but I made them large so that they could be scaled down a bit if you wanted that. It looks bad to scale something _up_.

I was thinking about updating the whole `assets/dungeon/`  directory, but since many of the SL dungeons won't be in a rotation for a long time, it seemed moot.

File: src/renderer/images.ts
Comment (tdaugaard) [reply src/renderer/images.ts:41]: Oversight - it shouldn't be there.

File: src/renderer/VideoButton.tsx
Comment (tdaugaard) [reply src/renderer/VideoButton.tsx:137]: > Best guess is to handle the case if M+ timers change in a patch, we can still be accurate for previous videos?

Precisely. Timers are occassionally, but rarely, adjusted in the beginning when a new dungeon enters the rotation. I'm neutral as to include this or not.

File: src/renderer/VideoButton.tsx
Comment (tdaugaard) [reply src/renderer/VideoButton.tsx:144]: `keystoneUpgradeLevel` - which is what it signifies - value 0 - 3.

File: src/renderer/VideoButton.tsx
Comment (tdaugaard) [reply src/renderer/VideoButton.tsx:152]: I have fixed it better and utilized the fact that `.push()` can take an arbitrary number of elements.

Comment (tdaugaard) [reply src/renderer/VideoButton.tsx:152]: I have fixed it better and utilized the fact that `.push()` can take an arbitrary number of elements.

File: src/renderer/rendererutils.ts
Comment (tdaugaard) [reply src/renderer/rendererutils.ts:35]: You mean why ChallengeMode need its own handling or why I'm testing for `undefined`?

**Comment by tdaugaard** [comment]
> What about timestamps in the test log (I commented on this in the other PR)?

I don't think we should design the app around the tests.

I've a log replay script in the works which can "replay" a given log file, adjusting the timestamps on the fly such that this wouldn't be a problem.

File: src/renderer/rendererutils.ts
Comment (aza547) [reply src/renderer/rendererutils.ts:35]: Both I suppose. Why isn't video.result === video.challengeMode.timed in this context?

**Comment by aza547** [comment]
> > What about timestamps in the test log (I commented on this in the other PR)?
> 
> I don't think we should design the app around the tests.
> 
> I've a log replay script in the works which can "replay" a given log file, adjusting the timestamps on the fly such that this wouldn't be a problem.

Maybe I wasn't clear - I'm just suggesting we tweak the timestamps in the dungeon.txt file which is part of the tests to make the seeking testable here. So a tweak to a test file, not a functional change.

File: src/renderer/rendererutils.ts
Comment (tdaugaard) [reply src/renderer/rendererutils.ts:35]: Because a dungeon can be completed, but not timed. If a dungeon isn't completed, `video.result` will be false, if it's completed by not timed `video.result = true` and `video.challengeMode.timed = false`.

**Comment by tdaugaard** [comment]
> Maybe I wasn't clear - I'm just suggesting we tweak the timestamps in the dungeon.txt file which is part of the tests to make the seeking testable here. So a tweak to a test file, not a functional change.

I probably misunderstood - I will make a new test file with adjusted timestamps. :)

File: src/renderer/rendererutils.ts
Comment (aza547) [reply src/renderer/rendererutils.ts:35]: Obvious really... thanks!

**Comment by aza547** [comment]
> > Maybe I wasn't clear - I'm just suggesting we tweak the timestamps in the dungeon.txt file which is part of the tests to make the seeking testable here. So a tweak to a test file, not a functional change.
> 
> I probably misunderstood - I will make a new test file with adjusted timestamps. :)

I had in mind some crappy interpolation to take number of lines in that test file, divide the 10 (seconds) by that and then re-create them all equally spaced throughout the 10s. I think that's pretty easy and good enough for test purposes (if you'd like to leave this to me, feel free).

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts:699]: > What about horribly bad runs with like 20 deaths, is that all going to appear in the context menu?

We will be better and die less! Just kidding, I get what you mean - especially on raid bosses where several players might die in close succession either because of mechanics of due to a wipe being called.

I _could_ make some kind of logic such that a player death would only count if there has been no other player deaths in X number of seconds, or simply ignore player deaths after a certain number.

For now, player deaths aren't exposed in the UI - I created the logic for it but ultimately decided to leave it out due to space issues - we'll talk about what to do with it after M+ is merged in, as discussed on Discord too. 

Does that sound OK?

> What about hunter feign deaths which causes a UNIT_DIED event to fire?

Good point! I can actually see here, that the very last value in the `UNIT_DIED` line indicates if the unit is "unconscious" (i.e. feigned death) so I can/will filter on that. Shouldn't have feign deaths included.

**Comment by tdaugaard** [comment]
> I thought video files should be 200x100 to be in-line with what we have already. I thought having large files causes the video button to display a corner of them. That being said it looks OK, so maybe you made some CSS changes that I've yet to see.

When using the loading screen images as the 'poster' for the video player, it wouldn't look nice to scale up a 200x100 picture. I've set `background-size: cover` for the `.videoButton.dungeon` selector.

> I don't think I'm a huge fan of the loading screen images (the WoW logo and text can overlap our text which I think looks bad.

Not sure what you mean with overlap - can you show me an example screenshot? Is it on the video button or as the video player 'poster' you don't like?

> The text is a different font/size to other buttons still. Can we make it consistent?

I've changed `.zone` text to be 16px (up from 13px) because the zone element is used for the key level (e.g. "++17"). `zone` class is used for both the dungeon name and the keystone upgrade & level.

I just want it to look good; I don't really care about whether it's 13px or 16px, but 13px seems to be too small - would it be OK to change it to 16px for everything?

> Logging is pretty ugly for the final metadata event, can we split so it's not a one line JSON string but a properly formatted JSON?

Agreed - done! 2 space indentation.

**Comment by aza547** [comment]
> > I thought video files should be 200x100 to be in-line with what we have already. I thought having large files causes the video button to display a corner of them. That being said it looks OK, so maybe you made some CSS changes that I've yet to see.
> 
> When using the loading screen images as the 'poster' for the video player, it wouldn't look nice to scale up a 200x100 picture. I've set `background-size: cover` for the `.videoButton.dungeon` selector.
> 
> > I don't think I'm a huge fan of the loading screen images (the WoW logo and text can overlap our text which I think looks bad.
> 
> Not sure what you mean with overlap - can you show me an example screenshot? Is it on the video button or as the video player 'poster' you don't like?



> 
> > The text is a different font/size to other buttons still. Can we make it consistent?
> 
> I've changed `.zone` text to be 16px (up from 13px) because the zone element is used for the key level (e.g. "++17"). `zone` class is used for both the dungeon name and the keystone upgrade & level.
> 
> I just want it to look good; I don't really care about whether it's 13px or 16px, but 13px seems to be too small - would it be OK to change it to 16px for everything?
> 
> > Logging is pretty ugly for the final metadata event, can we split so it's not a one line JSON string but a properly formatted JSON?
> 
> Agreed - done! 2 space indentation.

**Comment by aza547** [comment]
The text that I don't like is mostly the bits at the bottom FYI, the name and the timed/depleted indicator

File: src/main/logutils.ts
Comment (aza547) [reply src/main/logutils.ts:703]: Oh, that's awesome on the feign death thing. I had no idea that was available. That'd be great to include. 

Ultimately I think I don't care about this now I know it's not ending up in the context menu!

**Comment by tdaugaard** [comment]
> Agree, but I scaled down those pictures initially. No reason we can't get the high-res versions again for this in the future.

The pictures are large, and have to be large, because they are used as 'poster' image for the video player.
I personally like them and have no problem with the text overlapping, or rather overlaying, the background though.

Do you want me to:

1. Find non-loading screen looking images for the dungeons instead
2. Remove the 'poster' feature altogether and use non-loading screen looking low-res pictures
3. Keep the 'poster' feature and use different images for the video player 'poster' (current large ones) and the VideoButton, with the button using a low-res, non-loading screen looking, version of them

Let me know :)

**Comment by tdaugaard** [comment]
> The text that I don't like is mostly the bits at the bottom FYI, the name and the timed/depleted indicator

The size? It can be larger. I did make the player name smaller for M+ as I consider that information unimportant for a dungeon run, but acknowledge that it might be more important in the PvP categories.

**Comment by aza547** [comment]
> > The text that I don't like is mostly the bits at the bottom FYI, the name and the timed/depleted indicator
> 
> The size? It can be larger. I did make the player name smaller for M+ as I consider that information unimportant for a dungeon run, but acknowledge that it might be more important in the PvP categories.

Yeah the size. I think for people playing multiple chars this info is super useful. I'd also prefer it to stay the same size across different categories - if you could revert that :)

File: src/main/keystone.ts
Comment (tdaugaard) [reply src/main/keystone.ts]: `playerDeaths` has been moved to  `Metadata` type and the comment has been rewritten.

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts:43]: Commented about this elsewhere. There's a flag in UNIT_DIED combat log line (last one, number 9) that indicates this.

**Comment by aza547** [comment]
> > Agree, but I scaled down those pictures initially. No reason we can't get the high-res versions again for this in the future.
> 
> The pictures are large, and have to be large, because they are used as 'poster' image for the video player. I personally like them and have no problem with the text overlapping, or rather overlaying, the background though.
> 
> Do you want me to:
> 
> 1. Find non-loading screen looking images for the dungeons instead
> 2. Remove the 'poster' feature altogether and use non-loading screen looking low-res pictures
> 3. Keep the 'poster' feature and use different images for the video player 'poster' (current large ones) and the VideoButton, with the button using a low-res, non-loading screen looking, version of them
> 
> Let me know :)

I think my preference is non-loading screen images, no problem with the big images - agree with the reasoning for that. This isn't really a big deal. I probably want to re-style these buttons some point soon to look better as I think it's one of the weaker bits aesthetically.

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts:395]: > This is a bit ugly, can we split it out?

I'm generally against making variables that are only used in a single following comparison, but if that's what you want, I'll do it.

> Was also suprised to see no exit from this function here, don't you need to return to avoid processing a dungeon we don't recognise?

Yes and no. I was hoping to and planning on making it so that in the future when new dungeons are added, we wouldn't avoid recording and saving videos from unknown dungeons, encounters and such so that we could simply update the app and the details would show up after the recording had been saved.

`handleEncounterStartLine()` also doesn't care about the encounter existing or not.

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts:410]: _Theater of Pain_ comes to mind as one such dungeon where it was a regular thing to skip the initial 3 mobs, and bring the 4th on top of the boss - besides that I don't recall any dungeons where doing two bosses without any trash is possible; HOWEVER even if that was the case there would probably be a skip strategy that would be relevant to have a beginning timestamp of.

I could label the first segment as 'Intro' segment instead, if you want?

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts:458]: I don't know if it is, but the duration as taken here from argument 4 is the duration that the game considers for the keystone completion upgrade levels; I didn't want to take the risk that the (end - start) log line timestamps would be off by even 500 ms, as that _can_ mean the difference between timing a key or not timing it.

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts:451]: I think I mentioned that elsewhere;

`metadata.result` = keystone completed?
`metadata.challengeMode.timed` = if completed, was it timed?

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts:463]: Adding additional logging - you are right, but it is doing both things - ending the video segment _and_ the CM instance.
```typescript
console.debug("[ChallengeMode] Starting Challenge Mode instance")
```

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts:466]: Perhaps I could rename `VideoSegment` to `TimelineSegment`? I agree it could be confusing.

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts:482]: Yes we could!

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts:540]: I've done a lot of dungeons and I've never seen a boss being pulled _into_ a boss. Do you have an example?

**Comment by tdaugaard** [comment]
> Yeah the size. I think for people playing multiple chars this info is super useful. I'd also prefer it to stay the same size across different categories - if you could revert that :)

Gonna be a very tight fit though. I could change "Timed" and "Depleted" to "T" and "D" or something else; I don't know.

image

Example of a maximum length character name (12 characters) - it doesn't look great at all:

image

**Comment by aza547** [comment]
> > Yeah the size. I think for people playing multiple chars this info is super useful. I'd also prefer it to stay the same size across different categories - if you could revert that :)
> 
> Gonna be a very tight fit though. I could change "Timed" and "Depleted" to "T" and "D" or something else; I don't know.
> 
> image
> 
> Example of a maximum length character name (12 characters) - it doesn't look great at all:
> 
> image

Time / Depl would be OK imo?

**Comment by tdaugaard** [comment]
> Time / Depl would be OK imo?

Sounds good to me.

> Or even better +1, +2, +3 or "Depl"?

Actually, that's what the "+", "++", "+++" signifies in the keystone level - I thought it was a good way to show it as that is how some people write it when talking about expectations (in dungeon finder).

What do you prefer?

**Comment by aza547** [comment]
> > Time / Depl would be OK imo?
> 
> Sounds good to me.
> 
> > Or even better +1, +2, +3 or "Depl"?
> 
> Actually, that's what the "+", "++", "+++" signifies in the keystone level - I thought it was a good way to show it as that is how some people write it when talking about expectations (in dungeon finder).
> 
> What do you prefer?

Oh I see! I wasn't aware that was a convention (and maybe others wouldn't be either). I think I do prefer the +N style.

File: src/main/logutils.ts
Comment (aza547) [reply src/main/logutils.ts]: Ah, I am of the mindset that if it makes code readable then it's a good thing. I appreciate that's an opinion not a fact so I'm happy either way.

File: src/main/logutils.ts
Comment (aza547) [reply src/main/logutils.ts:413]: If it's just a case of labelling then I don't think I care too much -- If it's functionally a non-issue and the bottom line is we just occasionally label some strats wrongly then that seems like an easy fix to make when people complain about it rather than worry now.

File: src/main/logutils.ts
Comment (aza547) [reply src/main/logutils.ts:460]: OK, can we just say this in a comment? Sounds fine, but good to have for future understanding

File: src/main/logutils.ts
Comment (aza547) [reply src/main/logutils.ts:468]: I think I prefer videoSegment, but not overly fussed - would be happy with a comment here

File: src/main/logutils.ts
Comment (aza547) [reply src/main/logutils.ts:486]: (to be clear - fine if you don't do this now - just a comment for future!)

File: src/main/logutils.ts
Comment (aza547) [reply src/main/logutils.ts:547]: I had Taza'vesh in mind but I'm not sure as I basically never do M+.

 As per earlier happy enough if there is a few rough edges like this here.

**Comment by tdaugaard** [comment]
Honestly, I think this looks better too:

image

**Comment by tdaugaard** [comment]
I renamed everything related to "video segments" to be "timeline segments" to avoid confusing it for physical video data.

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts]: I split it. It's your project afterall :)

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts]: It's just labeling. I guess the "real fix" here is to make it so that if a timeline segment (video segment) is too short to have value, e.g. 10 seconds, then it'll be discarded and whatever comes next (like a boss) will take its place, but I think those refinements are better to keep for future improvements.

File: src/main/keystone.ts
Comment (tdaugaard) [reply src/main/keystone.ts]: > Could use an enum for this rather than assigning numbers.

An enum does not make sense here, in my opinion, since keystone upgrade levels is a numeric value by design and this method reflects that. I don't know how I should make it an enum even.

> I don't think this shows in the UI. Could we make the UI show +3, +2, +1 or depleted, if we already have the info?

Yes, that's been addressed in a newer commit regarding the look of the videobutton.

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts]: Especially Taza'vesh will have a skip before the first boss then, and although it's not a complicated skip by any stretch of imagination, I'd personally like to be able to review it if we screw it up :p 

The actual reason for having a first timeline segment before the first boss is just to simply be able to seek to 0:00 withuot having to use the video player seek bar.

File: src/main/logutils.ts
Comment (aza547) [reply src/main/logutils.ts]: Awesome - totally happy with that.

File: src/main/logutils.ts
Comment (tdaugaard) [reply src/main/logutils.ts]: I renamed the whole video segment thing because I did find it confusing myself but did't have a better alternative at the time. It's now called timeline segments. 🤷‍♂️

File: src/main/keystone.ts
Comment (aza547) [reply src/main/keystone.ts]: Yeah fair enough, happy with that.

File: src/main/logutils.ts
Comment (aza547) [reply src/main/logutils.ts]: Ah so it would just show like "trash" 0-30s but there would be no combat in that, just running into position? Actually I like that as it is then. I agree that's important stuff you want to see and be able to skip to!

**Review by aza547 — APPROVED** [review]
Thanks for all the hard work here. Excited to see this feature go live! 

I'll let you do the honors and hit merge :)

**Pull request merged**
